File: hashmaps_and_sets/groupAnagrams.go
package main

import (
	"fmt"
	"math/big"
)

/*

link leetcode problem -> https://leetcode.com/problems/group-anagrams

time complexity  -> O(n * k)

space complexity -> O(n * k)

where n is the length of strs and k length of strs[i]
*/

func groupAnagrams(strs []string) [][]string {
	result := [][]string{}

	primes := map[rune]int64{
		'a': 2, 'b': 3, 'c': 5, 'd': 7, 'e': 11, 'f': 13, 'g': 17, 'h': 19,
		'i': 23, 'j': 29, 'k': 31, 'l': 37, 'm': 41, 'n': 43, 'o': 47, 'p': 53,
		'q': 59, 'r': 61, 's': 67, 't': 71, 'u': 73, 'v': 79, 'w': 83, 'x': 89,
		'y': 97, 'z': 101,
	}


	mymap := make(map[string]int)
	positionArray := 0

	for _, word := range strs {
		prod := big.NewInt(1) 

	
		for _, char := range word {
			prod.Mul(prod, big.NewInt(primes[char]))
		}

		prodStr := prod.String()


		if pos, ok := mymap[prodStr]; !ok {
			mymap[prodStr] = positionArray
			result = append(result, []string{word})
			positionArray++
		} else {
			result[pos] = append(result[pos], word)
		}
	}

	return result
}





func main() {
	strs := []string{"aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa", "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaab"}
	fmt.Println(groupAnagrams(strs))
}

/*
97 -> a
98 -> b
101 -> e 33
110 -> n
116 -> t

*/

File: hashmaps_and_sets/group_anagrams.py
#Link to problem: https://leetcode.com/problems/group-anagrams/description/
#Time complexity:O(n * m), where n = length of strs and m = length of strs[i]
#Space complexity: O(n * m), where n = length of strs and m = length of strs[i]
                       
class Solution:
    def group_anagrams(self,strs:list[str]) -> list[list[str]]:
        if len(strs) == 1:
            return [[strs[0]]]
        
        map = {}
            
        for word in strs:
            counter = [0 for _ in range(26)]

            for letter in word:
                counter[ord(letter) - ord('a')] += 2
            
            value = '-'.join(str(v) for v in counter)
            if map.get(value) is None:
                map[value] = [word]
            else:
                 map[value].append(word)
    

        return list(map.values())

                                        
solution = Solution()
print(solution.group_anagrams(["eat","tea","tan","ate","nat","bat"]))
print(solution.group_anagrams([""]))
print(solution.group_anagrams(["a"]))
print(solution.group_anagrams(["bdddddddddd","bbbbbbbbbbc"]))